### PR TITLE
django-storages support

### LINKED ISSUE #13: Make sure it works with django-storages...
...especially with AWS S3

### PR DESCRIPTION
**Pull request by lifenautjoe**
Fixes #13

### CODE REVIEW

**Comment by lifenautjoe** [comment]
Managed to get the testing running by not using poetry as #18 .

However find the tests very confusing. Where are all the arguments to the test cases coming from? What is this magic? 🤔

**Comment by lifenautjoe** [comment]
Alright, its working and we're testing it in our project with an S3 backend and the tests here work.

What I did is if the storage doesn't support absolute paths, I read the file (in remote storages, this downloads the file) and write it to a temp file which is then discarded after exit.

I also handled the case where because its being used as a Mixin, it can be a simple file without a storage property. In that case we fall back to the `os.path.abspath(fieldfile.path)`.

The reason why I didn't cache files is that the video processing can happen in worker nodes, meaning any caching done in the first receiving serve, won't be accessible to the worker node.

Also because the processing is in an asynchronous process, we would have to somehow transfer the temporal file path to the sub process, which would then have to clean up. However in the case of a redeploy, while the task might still be scheduled, the file will no longer be there.

It's safer to re-download the file each time. Yes, more consuming but safer.

@bashu, here you go too.

This fixes #13

File: video_encoding/utils.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    else:
+        # Its a local file with no storage abstraction
+        try:
+            path = os.path.abspath(fieldfile.path)
+        except AttributeError:
+            path = os.path.abspath(fieldfile.name)
+
+        storage_local_path = path
+
+    return storage_local_path, local_temp_file
```
Comment (escaped) [inline video_encoding/utils.py:38]: I am wondering if we could make this implementation a bit more robust by using [contextlib.contextmanager](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager). Something like

```python
@contextlib.contextmanager
def get_local_path(fieldfile: File) -> str:
    if not hasattr(fieldfile, 'storage'): 
        storage = fieldfile.storage
        try:
            # Try to access with path
            yield storage.path(fieldfile.path)
        except (NotImplementedError, AttributeError):
            # Storage doesnt support absolute paths,
            # download file to a temp local dir
            with tempfile.NamedTemporaryFile(delete=False) as temp_file:
                storage_file = storage.open(fieldfile.name, 'r')

                temp_file.write(storage_file.read())

            try:
                yield temp_file.name
            finally:
                os.unlink(temp_file.name)
    else:
        # Its a local file with no storage abstraction
        try:
            yield os.path.abspath(fieldfile.path)
        except AttributeError:
            yield os.path.abspath(fieldfile.name)
```

with this, we could simply use it as a context manager and don't need to worry about closing and unlinking the temporary file.

```python
    def _get_video_info(self):
        """
        Returns basic information about the video as dictionary.
        """
        if not hasattr(self, '_info_cache'):
            encoding_backend = get_backend()
            with get_local_path(self) as local_path:
                self._info_cache = encoding_backend.get_media_info(path)
        return self._info_cache
```

What do you think about this? :)

File: video_encoding/utils.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import os
+import tempfile
+
+
+def get_fieldfile_local_path(fieldfile):
```
Comment (escaped) [inline video_encoding/utils.py:5]: Would be great, if we could add a test for this. I know the current folder structure is a bit weird, but I would suggest adding `test_proj/video_encoding/test_utils.py` for now. I'll refactor the tests into a separate directory later.

**Review by escaped — CHANGES_REQUESTED** [review]
Nice work! Thank you for looking into this and providing a PR. I added some suggestions and was hoping that you are still interested in working on this PR. If not, then I will gladly take over.

As you might noticed, I restructured CI and we are using github actions instead of travis now. While doing this I introduced some new tools like `isort` and `black`, which are now mandatory. Could you please update your PR to make them happy? :)

```sh
poetry install
poetry run pre-commit install
poetry run pre-commit run -a  # fix linting issues locally
```

**Comment by lifenautjoe** [comment]
Hi @escaped , sorry, haven't had much time to look into this. If you have some time, please feel free to pick it up.

Thank you!

**Comment by escaped** [comment]
Closing in favour of #26. Thank you for doing the initial work :)

**Pull request closed without merging**